### ISSUE TITLE
valid-jsdoc: validate that scalar types are written in lowercase

### ISSUE THREAD

**Issue by janmarek** [eslint/eslint#3056]
I personally write number, string, boolean and function types in lowercase and Object and Array with first capital letter. I would appreciate if it could be validated.

    @param {number} num
    @return {Object} 
    ...

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by gyandeeps** [comment]
@eslint/eslint-team thoughts on this?

**Comment by nzakas** [comment]
I like it!

**Comment by ilyavolodin** [comment]
I think this needs to be configurable in some way. Some people would use `object` and some would use `Object`.

**Comment by nzakas** [comment]
Object would not be checked by this option.

I do think it needs to be configurable. For instance, i use Function not function because functions are objects.

**Comment by janmarek** [comment]
I agree it should be configurable.

**Comment by jaydenseric** [comment]
This would be super useful, because I am constantly wondering what case to use for certain types. Actual JSDoc documentation is inconsistent in their examples.

**Comment by platinumazure** [comment]
I don't know if the JSDoc spec cares one way or the other, but even for "number" I could see arguments both ways (constructor is "Number", typeof output is "number"). So unless JSDoc itself proscribes one way or the other, I don't think ESLint should (that is, it should allow for both ways via configuration).

**Comment by nzakas** [comment]
@platinumazure it doesn't matter what JSDoc does, we can still apply additional constraints. We just need to make them configurable.

**Comment by platinumazure** [comment]
@nzakas Agreed. What I was saying is, we shouldn't apply un-configurable
constraints unless JSDoc clearly proscribes said constraints.

**Comment by nzakas** [comment]
Yup. That's what I said here: https://github.com/eslint/eslint/issues/3056#issuecomment-130733039

**Comment by gyandeeps** [comment]
I am planning to work on this. Any recommendation on how the option would be configured?

**Option 1**
Define preferences for the standard types and they will have defaults.
```json
{
    "string": "string",
    "object": "Object",
    "number": "Number"
}
```

**Option 2**
Set it to `true` to make it start with capital letter
```json
{
    "string": false,
    "object": true,
    "number": true
}
```

**Comment by nzakas** [comment]
I think option 1 is right, maybe called preferType?

**Comment by gyandeeps** [comment]
Here is the default I am thinking of

```js
var PREFER_TYPE_DEFAULTS = {
    "string": "string",
    "object": "object",
    "number": "Number",
    "function": "Function",
    "boolean": "boolean",
    "int": "Int",
    "array": "Array"
};
```

Please suggests corrections for the above.

Should we start checking types using the defaults for this rule or would this be a opt-in feature where only the option specified by the user would be checked?

CC @nzakas @eslint/eslint-team

**Comment by mysticatea** [comment]
I like lower cases for types which are get by typeof operator -- boolean, number, string, object, function, symbol, and undefined.

**Comment by ilyavolodin** [comment]
@gyandeeps Why not just pass in an array of they values as you'd like them to appear in JSDoc? Something like this:
```[2, { preferType: ["string", "object", "Number", "Function", "boolean", "Int", "Array"]} ]```

**Comment by nzakas** [comment]
Anything that isn't an object should be lowercase, so "boolean" and "int"

"Object" and "Function" should be uppercase because they are objects. 

@ilyavolodin what does an array of strings mean in this case?

**Comment by ilyavolodin** [comment]
Array of string the way they should be seen in JSDoc comments. If array contains "Object" that means anytime "object" is found, it should warn. Simpler then duplicating every word i.e. `{ "object": "Object" }`

**Comment by nzakas** [comment]
You're talking about a whitelist, which I don't believe is feasible. There's no way users will constantly update this list for every type they expect to see (which includes custom classes). The blacklist approach is more reasonable for this use case.

**Comment by jaydenseric** [comment]
@ilyavolodin I like the simple array idea. @nzakas It could still be a blacklist, by doing a [case-insensitive lookup](http://stackoverflow.com/questions/24718349/how-do-i-make-array-indexof-case-insensitive). If found, the result is then compared to the test value case-sensitive.

**Comment by ilyavolodin** [comment]
@jaydenseric Right, that's exactly what I mean. Do case insensitive lookup and then fail if the case-sensitive lookup fails.

**Comment by nzakas** [comment]
Not sure I follow, can you give an example?

**Comment by ilyavolodin** [comment]
Given a jsDoc like this:
```js
/**
 * @param {string} string The string to parse.
 */
```
And a configuration like this:
```js
[2, { preferType: ["String", "object", "Number", "Function", "boolean", "Int", "Array"]} ]
```

The code can parse argument type from the param and do a case-insensitive lookup from the `preferType` array. It will get location 0, then it can do a case-sensitive comparison of the type to `preferType[0]` and if they are not the same, error out.

**Comment by janmarek** [comment]
@ilyavolodin This configuration format looks good to me.

**Comment by jaydenseric** [comment]
This might be a good guide for capitalization defaults: [*Google JavaScript Style Guide*, *JavaScript Types* section](https://google.github.io/styleguide/javascriptguide.xml?showone=JavaScript_Types#JavaScript_Types). In particular, look at the *Type Example* column in the table under *Types in JavaScript*.

**Comment by janmarek** [comment]
Number/String objects are very rare. I've personally never used any of them. Validation with exception for these types would be useless.

**Comment by nzakas** [comment]
@ilyavolodin I'm still having a hard time with this, and the automatic conversion to all lowercase seems a bit magical. I still think being able to say "if you see x then it should be y" is more extensible and explicit than the single array, and follows conventions we already use.

**Comment by ilyavolodin** [comment]
@nzakas I totally agree that it's more explicit, it's just that you need to repeat yourself a lot, and seems a bit redundant. But I'm fine either way.

**Comment by tschaub** [comment]
One nuance specific to the Closure Compiler is that `Function` and `function(number)` have specific meaning: the former is any function, the latter is a function that takes a single `number` argument.

In addition, in this strict types environment, the following are valid values for the given types:

 * `number` : `10`, `NaN`
 * `Number` : `new Number(10)`, `null`
 * `string` : `'foo'`
 * `String` : `new String('foo')`, `null`
 * `boolean` : `true`
 * `Boolean` : `new Boolean(true)`, `null`
 * `Object` : `{}`, `null`
 * `Array` : `[]`, `null`

So I don't think the `preferType` in 563d97d750fb404b22945c6d9640145dc59ca850 is going to work in this situation.

I think it would be unfortunate if ESLint's valid-jsdoc rule were strict in a way that made it incompatible with type checking.

**Comment by tschaub** [comment]
> So I don't think the preferType in 563d97d is going to work in this situation.

Actually, I take that back.  This would work:

```js
preferTypes: {
  array: 'Array',
  object: 'Object,
}
```

The linter would warn about typos in those two cases.
